### ISSUE TITLE
Allow ignoring part of history when calculating version

### ISSUE THREAD

**Issue by JakeGinnivan** [GitTools/GitVersion#538]
Based on discussions from #470 we reached a consensus that you should be able to ignore before a time period. This can be a configuration value:

`ignore-commits-before: 2015/01/01` 

This allows when repositories have bad commit messages or something else which is causing a high version number (see #443).

```
ignore-shas:
    sha1
    sha2
```

could be another option, the logs tell you what commits different versions are coming from, so it would be copied straight out of the logs to ignore that particular commit. Having both would not be bad.

**Comment by asbjornu** [comment]
If you go for a date, please use RFC 3999 (subset of ISO 8601) for the format. That means the above example should be "2015-01-01". :)

**Comment by asbjornu** [comment]
And perhaps it would be better for extensibility to group everything under an `ignore` section, like so?
```yaml
ignore:
    shas: b6c0c9fda88830ebcd563e500a5a7da5a1658e98 6c19c7c219ecf8dbc468042baefa73a1b213e8b1
    commits-before: 2015-01-01
```
Grouping it like this would make it easy to add `commits-between` in the future, for instance.

What should the syntax for several SHAs be, by the way? Separated by space or linefeed and indented?

**Comment by dmitriy-konovalov** [comment]
I ran into problem with ip address in commit message too (as described in #443). Is there any plans implementing history part ignoring in near future?

**Comment by orjan** [comment]
Dmitry, do you mind publishing the commit message and the GV version, so we can create a failing test?

-- 
Örjan Sjöholm

**Comment by dmitriy-konovalov** [comment]
Sure!
Git commit message example 1:
'''
Merge branch 'master' of http://212.248.89.56:8082/r/asu_tk/p_sd
'''

Git commit message example 2:
'''
Merge branch 'master' of http://172.16.3.10:8082/r/asu_tk/p_sd
'''

GitVersion log output:
'''
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:51] Working directory: C:\Users\d.konovalov\Documents\C#\p_sd
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:55] Project root is: C:\Users\d.konovalov\Documents\C#\p_sd
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:66] Using latest commit on specified branch
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:68] Running against branch: develop (d1a531eec15911111d18783e2550f9c193e6acac)
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:70] Begin: Calculating base versions
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:73] Fallback base version: 0.1.0 with commit count source b08107bc350cba5ae0757d686de7972c212690f9
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:79] Merge message 'Finish 1.1.0': 1.1.0 with commit count source 8b298b8e6de1cf755b2193e4acb082e33a67b666
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:79] Merge message 'Merge branch 'master' of http://212.248.89.56:8082/r/asu_tk/p_sd': 212.248.89+56 with commit count source 77e69a601a9231a48cd1f052f78a452037444404
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:80] Merge message 'Merge branch 'master' of http://172.16.3.10:8082/r/asu_tk/p_sd': 172.16.3+10 with commit count source c3af15bb86ba5fa8d8371fed50e5cc2df9867b85
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:81] Merge message 'Merge branch 'master' of http://172.16.3.10:8082/r/asu_tk/p_sd': 172.16.3+10 with commit count source 3c466eb0d952f5dcb9ed4aad5c78499c056bfaaf
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:82] Merge message 'Merge branch 'master' of http://172.16.3.10:8082/r/asu_tk/p_sd': 172.16.3+10 with commit count source b28df3142d33c94fbaed33cc7b31e2e73b66823d
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:83] Merge message 'Merge branch 'master' of http://172.16.3.10:8082/r/asu_tk/p_sd': 172.16.3+10 with commit count source e6959fd920fd8a5417dfd72c7f033a2da0baa635
2015-09-18 21:19:45		  INFO [09/18/15 21:19:45:86] Base version used: Merge message 'Merge branch 'master' of http://212.248.89.56:8082/r/asu_tk/p_sd': 212.248.89+56 with commit count source 77e69a601a9231a48cd1f052f78a452037444404
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:87] End: Calculating base versions (Took: 164.02ms)
2015-09-18 21:19:45		INFO [09/18/15 21:19:45:90] 43 commits found between 77e69a601a9231a48cd1f052f78a452037444404 and d1a531eec15911111d18783e2550f9c193e6acac
'''

Result:
'''
{
  "Major":212,
  "Minor":249,
  "Patch":"0",
  "PreReleaseTag":"unstable.43",
  "PreReleaseTagWithDash":"-unstable.43",
  "BuildMetaData":"",
  "BuildMetaDataPadded":"",
  "FullBuildMetaData":"Branch.develop.Sha.d1a531eec15911111d18783e2550f9c193e6acac",
  "MajorMinorPatch":"212.249.0",
  "SemVer":"212.249.0-unstable.43",
  "LegacySemVer":"212.249.0-unstable43",
  "LegacySemVerPadded":"212.249.0-unstable0043",
  "AssemblySemVer":"212.249.0.0",
  "FullSemVer":"212.249.0-unstable.43",
  "InformationalVersion":"212.249.0-unstable.43+Branch.develop.Sha.d1a531eec15911111d18783e2550f9c193e6acac",
  "BranchName":"develop",
  "Sha":"d1a531eec15911111d18783e2550f9c193e6acac",
  "NuGetVersionV2":"212.249.0-unstable0043",
  "NuGetVersion":"212.249.0-unstable0043",
  "CommitDate":"2015-09-18"
}
'''

**Comment by mkoertgen** [comment]
Issues commente in https://github.com/GitTools/GitVersion/issues/538#issuecomment-141526150 are fixed with #661 .

**Comment by dmitriy-konovalov** [comment]
Thanks!

**Comment by Tungsten78** [comment]
An errant merge comment is plaguing our versioning. I will submit a PR soon for this one.

**Comment by Tungsten78** [comment]
See PR #851

**Comment by Tungsten78** [comment]
@JakeGinnivan   This one can be closed.

**Comment by Tungsten78** [comment]
This issue was resolved in [v3.5.3](https://github.com/GitTools/GitVersion/releases/tag/v3.5.3)

**Comment by JakeGinnivan** [comment]
Great, thanks
